tests/pytests/l3_dump_test.py

```diff
@@ -0,0 +1,545 @@
+    # grep through source file and generate the expected msgs and
+    # unpacked LOC-IDs in a list.
+    # -----------------------------------------------------------------------
+    (exp_loc_list, exp_msg_list) = build_exp_loc_msg_list(usecase_prog_dir,
+                                                          src_filename)
+    assert msg_list == exp_msg_list
+
+    # LOC-ID will be generated and will be unpacked only if env-var is ON.
+    if os.getenv('L3_LOC_ENABLED') == L3_LOC_DEFAULT:
+        assert loc_list == exp_loc_list
+    else:
+        assert verify_loc_field_is_empty(loc_list) is True
+
+    exp_arg1_list = [ 1,
+                      int('0xdeadbabe', 16), int('0xbeefabcd', 16),
+                      10, 20
+                     ]
+    assert arg1_list == exp_arg1_list
+
+    exp_arg2_list = [ 2, 1024, 0, int('0xdeadbeef', 16), int('0xbeefbabe', 16) ]
+    assert arg2_list == exp_arg2_list
+
+    verify_rv = verify_output_lists(nentries, len(exp_msg_list),
+                                    tid_list, loc_list,
+                                    msg_list,
+                                    arg1_list, arg2_list)
+    assert verify_rv is True
+    return verify_rv
+
+# #############################################################################
+def verify_file_exists(dirname:str, filename:str) -> bool:
+    """
+    Check that the expected [generated] file exists.
+    """
+    return os.path.exists(dirname + '/' + filename)
+
+# #############################################################################
+def exec_make(cmdargs:list, extra_env:dict = None) -> bool:
+    """Execute `make` command with specified arguments."""
+    if extra_env is None:
+        extra_env = {}
+    try:
+        # Need to run `make` from the dir where Makefile lives.
+        # Add-in user-specified env-vars to existing env of process.
+        result = sp.run(cmdargs, text=True, check=True, capture_output=True,
+                        cwd = L3RootDir,
+                        env={**os.environ, **extra_env}
+                        )
+    except sp.CalledProcessError as exc:
+        print("sp.run() Status: FAIL, rc =", exc.returncode,
+              "\nargs =", exc.args,
+              "\nstdout =", exc.stdout,
+              "\nstderr =", exc.stderr)
+        return False
+    for line in str(result).split('\\n'):
+        print(line)
+    return True
+
+# #############################################################################
+def exec_binary(cmdargs:list):
+    """Execute a binary, with args, and print output to stdout."""
+
+    with sp.Popen(cmdargs,
+                  stdout=sp.PIPE,
+                  stderr=sp.PIPE, text=True) as section:
+        _stdout, _stderr = section.communicate()
+
+    print('\nExec:', cmdargs)
+    print(_stdout)
+    return True
+
+# #############################################################################
+def build_exp_loc_msg_list(prog_dir:str, main_file:str) -> (list, list):
```

Comment:
This is a **KEY** test-support routine.

We make testing reliable by `grep`-ping actual L3-logging APIs from the source-code, so that the line numbers are automatically, always, correct.

The decoded LOC-ID string format is re-constructed by this method.